File: src/CCVTAC.Console/PostProcessing/PostProcessing.cs
using System.IO;
using System.Text.Json;
using System.Text.RegularExpressions;
using CCVTAC.Console.IoUtilities;
using CCVTAC.Console.PostProcessing.Tagging;
using UserSettings = CCVTAC.FSharp.Settings.UserSettings;
using static CCVTAC.FSharp.Downloading;

namespace CCVTAC.Console.PostProcessing;

internal static partial class PostProcessor
{
    internal static readonly string[] AudioExtensions =
        [".aac", ".alac", ".flac", ".m4a", ".mp3", ".ogg", ".vorbis", ".opus", ".wav"];

    internal static void Run(UserSettings settings, MediaType mediaType, Printer printer)
    {
        Watch watch = new();
        string workingDirectory = settings.WorkingDirectory;

        printer.Info("Starting post-processing...");

        var taggingSetsResult = GenerateTaggingSets(workingDirectory);
        if (taggingSetsResult.IsFailed)
        {
            printer.Error("No tagging sets were generated, so tagging cannot be done.");
            return;
        }
        var taggingSets = taggingSetsResult.Value;

        var collectionJsonResult = GetCollectionJson(workingDirectory);
        CollectionMetadata? collectionJson;
        if (collectionJsonResult.IsFailed)
        {
            printer.Debug($"No playlist or channel metadata found: {collectionJsonResult.Errors.First().Message}");
            collectionJson = null;
        }
        else
        {
            printer.Debug("Found playlist/channel metadata.");
            collectionJson = collectionJsonResult.Value;
        }

        if (settings.EmbedImages)
        {
            ImageProcessor.Run(workingDirectory, printer);
        }

        var tagResult = Tagger.Run(settings, taggingSets, collectionJson, mediaType, printer);
        if (tagResult.IsSuccess)
        {
            printer.Info(tagResult.Value);

            Renamer.Run(settings, workingDirectory, printer);

            Mover.Run(taggingSets, collectionJson, settings, true, printer);

            var taggingSetFileNames = taggingSets.SelectMany(set => set.AllFiles).ToList();
            Deleter.Run(taggingSetFileNames, collectionJson, workingDirectory, printer);

            var leftoverFilesResult = Directories.WarnIfAnyFiles(workingDirectory, 20);
            if (leftoverFilesResult.IsFailed)
            {
                printer.FirstError(leftoverFilesResult);

                printer.Info("Will delete the remaining files...");
                var deleteResult = Directories.DeleteAllFiles(workingDirectory, 20);
                if (deleteResult.IsSuccess)
                {
                    printer.Info($"{deleteResult.Value} file(s) deleted.");
                }
                else
                {
                    printer.FirstError(deleteResult);
                }
            }
        }
        else
        {
            printer.Errors("Tagging error(s) preventing further post-processing: ", tagResult);
        }

        printer.Info($"Post-processing done in {watch.ElapsedFriendly}.");
    }

    private static Result<CollectionMetadata> GetCollectionJson(string workingDirectory)
    {
        try
        {
            var fileNames = Directory.GetFiles(workingDirectory)
                                     .Where(f => CollectionMetadataRegex().IsMatch(f))
                                     .ToImmutableHashSet();

            if (fileNames.Count == 0)
            {
                return Result.Fail("No relevant files found.");
            }

            if (fileNames.Count > 1)
            {
                return Result.Fail("Unexpectedly found more than one relevant file, so none will be processed.");
            }

            string fileName = fileNames.Single();
            string json = File.ReadAllText(fileName);
            var collectionData = JsonSerializer.Deserialize<CollectionMetadata>(json);

            return Result.Ok(collectionData);
        }
        catch (Exception ex)
        {
            return Result.Fail($"{ex.Message}");
        }
    }

    private static Result<ImmutableList<TaggingSet>> GenerateTaggingSets(string directory)
    {
        try
        {
            string[] files = Directory.GetFiles(directory);
            var taggingSets = TaggingSet.CreateSets(files);

            return taggingSets.Any()
                ? Result.Ok(taggingSets)
                : Result.Fail($"No tagging sets were created using working directory \"{directory}\".");
        }
        catch (DirectoryNotFoundException)
        {
            return Result.Fail($"Directory \"{directory}\" does not exist.");
        }
        catch (Exception ex)
        {
            return Result.Fail($"Error reading working directory files: {ex.Message}");
        }
    }

    /// <summary>
    /// A regular expression that detects metadata files for collections.
    /// </summary>
    [GeneratedRegex("""(?<=\[)[\w\-]{17,}(?=\]\.info.json)""")]
    private static partial Regex CollectionMetadataRegex();
}

File: src/CCVTAC.FSharp/PostProcessing/PostProcessing.fs
namespace CCVTAC.Console.PostProcessing

open CCVTAC.Console
open CCVTAC.Console.IoUtilities
open CCVTAC.Console.PostProcessing.Tagging
open CCVTAC.Console.Settings.Settings
open System.IO
open System.Linq
open System.Text.Json
open System.Text.RegularExpressions
open Startwatch.Library
open TaggingSets

module PostProcessor =

    let private collectionMetadataRegex =
        Regex(@"(?<=\[)[\w\-]{17,}(?=\]\.info.json)", RegexOptions.Compiled)

    let private isCollectionMetadataMatch (path: string) : bool =
        collectionMetadataRegex.IsMatch path

    let private getCollectionJson workingDirectoryName : Result<CollectionMetadata, string> =
        try
            let fileNames =
                Directory.GetFiles workingDirectoryName
                |> Seq.filter isCollectionMetadataMatch
                |> Set.ofSeq

            if Seq.isEmpty fileNames then
                Error "No relevant files found."
            elif Seq.hasMultiple fileNames then
                Error "Unexpectedly found more than one relevant file, so none will be processed."
            else
                let fileName = fileNames.Single()
                let json = File.ReadAllText fileName
                match JsonSerializer.Deserialize<CollectionMetadata> json with
                | Null -> Error $"Deserialized collection metadata for \"%s{fileName}\" was null."
                | NonNull parsedData -> Ok parsedData
        with
        | ex -> Error ex.Message

    let private generateTaggingSets dir : Result<TaggingSet list, string> =
        try
            let taggingSets = createSets <| Directory.GetFiles dir
            if List.isEmpty taggingSets
            then Error $"No tagging sets were created using files in working directory \"%s{dir}\". Are all file extensions correct?"
            else Ok taggingSets
        with exn ->
            Error $"Error reading working files in \"{dir}\": %s{exn.Message}"

    let run settings mediaType (printer: Printer) : unit =
        let watch = Watch()
        let workingDirectory = settings.WorkingDirectory

        printer.Info "Starting post-processing..."

        match generateTaggingSets workingDirectory with
        | Error _ ->
            printer.Error $"No tagging sets were generated for directory {workingDirectory}, so tagging cannot be done."
        | Ok taggingSets ->
            let collectionJson =
                match getCollectionJson workingDirectory with
                | Error e ->
                    printer.Debug $"No playlist or channel metadata found: %s{e}"
                    None
                | Ok cm ->
                    printer.Debug "Found playlist/channel metadata."
                    Some cm

            if settings.EmbedImages then
                ImageProcessor.run workingDirectory printer

            match Tagger.run settings taggingSets collectionJson mediaType printer with
            | Ok msg ->
                printer.Info msg
                Renamer.run settings workingDirectory printer
                Mover.run taggingSets collectionJson settings true printer

                let allTaggingSetFiles = taggingSets |> List.collect allFiles
                Deleter.run allTaggingSetFiles collectionJson workingDirectory printer

                match Directories.warnIfAnyFiles 20 workingDirectory with
                | Ok _ -> ()
                | Error err ->
                    printer.Error err
                    printer.Info "Will delete the remaining files..."
                    match Directories.deleteAllFiles workingDirectory with
                    | Ok results -> Directories.printDeletionResults printer results
                    | Error e -> printer.Error e
            | Error e ->
                printer.Error($"Tagging error(s) preventing further post-processing: {e}")

        printer.Info $"Post-processing done in %s{watch.ElapsedFriendly}."
